t/21-needles.t

```diff
@@ -168,4 +180,82 @@ subtest 'handling relative paths in update_needle' => sub {
+        is($needle_dirs, undef, 'needles directory determined');
+        is($needles_var, undef, 'needles var determined');
+    };
+
+    #subtest 'checkout_needles_sha = yes' => sub {
+    #    $controller->app->config->{'scm git'} = {checkout_needles_sha => 'yes'};
+    #    my ($needle_dirs, $needles_var) = $controller->_determine_needles_dir_for_job($job);
+    #    like($needle_dirs, qr|t/data/openqa/share/tests/fedora/needles$|, 'needles dirs is defined');
+    #    is($needles_var->value, 'https://something#fragment', 'needles var is defined');
+    #};
+};
+
+setup_fullstack_temp_dir('21-needles');
+
+subtest 'controller->_create_tmpdir_for_needles_refspec' => sub {
+    my $controller = OpenQA::WebAPI::Controller::Step->new;
+    $controller->app($t->app);
+    $controller->app->config->{'scm git'} = {checkout_needles_sha => 'yes'};
+
+    subtest 'without fragment and needles_git_hash in vars' => sub {
+        local $settings{CASEDIR} = 'https://something';
+        my $job = $schema->resultset('Jobs')->create_from_settings(\%settings);
+        path($job->needle_dir)->make_path;
+        $job->create_result_dir;
+
+        my $needle_refs = $controller->_create_tmpdir_for_needles_refspec($job);
+
+        is($needle_refs, undef, 'needles refs cannot be defined without fragment or vars.json');
+    };
+
+    subtest 'with vars' => sub {
+        local $settings{CASEDIR} = 'https://something';
+        my $job = $schema->resultset('Jobs')->create_from_settings(\%settings);
+        path($job->needle_dir)->make_path;
+        my $testresult_dir = $job->create_result_dir;
+        my $vars_json = path($testresult_dir, 'vars.json');
+        my $vars_json_content = '{"NEEDLES_GIT_HASH": "0123456789abcdef0123456789abcdef01234567"}';
+        $vars_json->spew($vars_json_content);
+
+        my $needle_refs = $controller->_create_tmpdir_for_needles_refspec($job);
+
+        is($needle_refs, '0123456789abcdef0123456789abcdef01234567', 'needles refs is defined with vars.json');
+    };
+};
+
+#subtest 'needledir set by jsonfile_in_temp_dir' => sub {
+#    my $needles_dir = path($ENV{OPENQA_BASEDIR}, 'openqa/webui/cache/needle-refs');
+#    path($needles_dir)->make_path;
+#    my $pngfile = path($needles_dir . '/needle.png');
+#    my $jsonfile = path($needles_dir . '/needle.json');
+#    $jsonfile->spew('{"area": [{"x" : 123, "y" : 456}]}');
+#    my $controller = OpenQA::WebAPI::Controller::File->new;
+#    $controller->app($t->app);
+#    $controller->param(name => $pngfile);
+#    $controller->param(version => $settings{VERSION});
+#    $controller->param(distri => $settings{DISTRI});
+#    $controller->param(jsonfile => $jsonfile);
+#
+#    $controller->needle;
+#
+#    $t->get('/needles/' . $settings{DISTRI} . '/needle.png')
+#      ->status_is(200);
+#    my $res = $t->name('needle_file')->to('file#needle');
+#    is($res, '', 'should fail');
```

Comment:
I leave here the comment by @Martchus done in the pr I was working on https://github.com/os-autoinst/openQA/pull/5175#discussion_r1954384059